pyccel/ast/basic.py

```diff
@@ -307,11 +308,20 @@ class PyccelAstNode(Basic):
     """Class from which all nodes containing objects inherit
     """
     stage      = None
-    _shape     = None
-    _rank      = None
-    _dtype     = None
-    _precision = None
-    _order     = None
+    __slots__ = ('_shape', '_rank', '_dtype', '_precision', '_order')
+
+    def __init__(self):
+        super().__init__()
+        if not hasattr(self, '_shape'):
+            self._shape = None
+        if not hasattr(self, '_rank'):
+            self._rank = None
+        if not hasattr(self, '_dtype'):
+            self._dtype = None
+        if not hasattr(self, '_precision'):
+            self._precision = None
+        if not hasattr(self, '_order'):
+            self._order = None
```

Comment:
I wonder if we really want every `PyccelAstNode` object to have all these slots. Couldn't we achieve the same behavior using the properties directly? For example

```python
class PyccelAstNode(Basic):
    stage = None
    __slots__ = ()

    # No __init__() function!

    @property
    def shape(self):
        return None

    @property
    def dtype(self):
        return None

    @property
    def precision(self):
        return None

    @property
    def order(self):
        return None
```

By the way, I don't recall the reason for having two separate classes `Basic` and `PyccelAstNode`. Can you remind me please?

Reply:
It is also possible to declare class-wide default values as before and let the subclasses declare only the slots that are needed:

```python
class PyccelAstNode(Basic):

    stage = None

    _shape     = None
    _rank      = None
    _dtype     = None
    _precision = None
    _order     = None

    __slots__ = ()  # fundamental!


class PythonInt(PyccelAstNode):
    _dtype = 'int'

    __slots__ = ('_precision',)
```

This would also allow us to remove properties in the future, if we need to further optimize our code.

Reply:
The names are poorly chosen. I had a discussion about this with @saidctb recently. `Basic` ought to be named `PyccelAstNode` and `PyccelAstNode` ought to be named something else, although I'm not sure what.
In the current state:
`Basic` - All classes in the pyccel AST inherit from this class. It takes care of the fst and the tree
`PyccelAstNode` - This represents an object that can be used in an `Assign` expression. Every object that inherits from this class should set:
- dtype
- precision
- shape
- rank
In addition if the rank > 1 then it should also set:
- order

I think there may be 1 or 2 classes which don't do this, hence the need for `if hasattr` but probably these classes either have bugs or ought to inherit from `Basic` instead. Maybe we should create an issue to clean up these classes

Reply:
> I wonder if we really want every `PyccelAstNode` object to have all these slots

If it inherits from `PyccelAstNode` and it doesn't need these slots then it probably shouldn't inherit from `PyccelAstNode`

Reply:
> Couldn't we achieve the same behavior using the properties directly?

Yes, but then we would need to override the property in lots of places. If we do this then there is no point in this class

Reply:
> It is also possible to declare class-wide default values as before and let the subclasses declare only the slots that are needed

This is possible, but again it will lead to a lot of duplication

Reply:
> > I wonder if we really want every `PyccelAstNode` object to have all these slots
> 
> If it inherits from `PyccelAstNode` and it doesn't need these slots then it probably shouldn't inherit from `PyccelAstNode`

I don't necessarily agree on this. A slot is something that changes from one object to another of the same class. So many subclasses do not need those slots, but just a class-wide value for those properties.

Reply:
Yes, ok. It doesn't need the slots, it just needs the attributes. I sent the first message without having seen your second message so I thought you were thinking of objects with dtype = None

Reply:
> It is also possible to declare class-wide default values as before and let the subclasses declare only the slots that are needed

This should be possible I think.

Reply:
My basic point is this: since `__slots__` are an optimization, we should be careful to use them the best we can, otherwise we are not going to improve anything and all our efforts will be in vain.

Reply:
```
class PyccelAstNode(Basic):

    stage = None

    __slots__ = ()  # fundamental!

    @property
    def shape(self):
        return self._shape #pylint : disable=no-member
```

all subclasses must either define a class property or a slot for:
- dtype
- precision
- shape
- rank
- order